File: packages/wazeko-protocol/src/encoder.ts
import { ProtocolNode } from "./node.js";
import { getTokenIndex } from "./tokens.js";

export const TAG_LIST_EMPTY = 0x00;
export const TAG_LIST_8 = 0x01;
export const TAG_LIST_16 = 0x02;
export const TAG_BINARY_8 = 0x03;
export const TAG_BINARY_32 = 0x04;
export const TAG_STRING = 0x05;
export const TAG_TOKEN = 0x06;

export class BinaryEncoder {
  private buffer: number[] = [];

  encode(node: ProtocolNode): Uint8Array {
    this.buffer = [];
    this.writeNode(node);
    return new Uint8Array(this.buffer);
  }

  private writeNode(node: ProtocolNode): void {
    const hasContent = node.content !== null && node.content !== undefined;
    const attrEntries = Object.entries(node.attrs);
    const listSize = 1 + attrEntries.length * 2 + (hasContent ? 1 : 0);

    this.writeListHeader(listSize);
    this.writeString(node.tag);

    for (const [key, value] of attrEntries) {
      this.writeString(key);
      this.writeString(value);
    }

    if (hasContent) {
      if (node.content instanceof Uint8Array) {
        this.writeBinary(node.content);
      } else if (Array.isArray(node.content)) {
        this.writeListHeader(node.content.length);
        for (const child of node.content) {
          this.writeNode(child);
        }
      }
    }
  }

  private writeListHeader(size: number): void {
    if (size === 0) {
      this.buffer.push(TAG_LIST_EMPTY);
    } else if (size < 256) {
      this.buffer.push(TAG_LIST_8, size);
    } else if (size < 65536) {
      this.buffer.push(TAG_LIST_16, (size >> 8) & 0xff, size & 0xff);
    } else {
      throw new Error(`List size exceeds 65535: ${size}`);
    }
  }

  private writeString(text: string): void {
    const tokenIndex = getTokenIndex(text);
    if (tokenIndex !== undefined && tokenIndex < 256) {
      this.buffer.push(TAG_TOKEN, tokenIndex);
    } else {
      const bytes = new TextEncoder().encode(text);
      const len = bytes.length;
      if (len > 65535) {
        throw new Error("String length exceeds 65535");
      }
      this.buffer.push(TAG_STRING, (len >> 8) & 0xff, len & 0xff);
      for (let i = 0; i < bytes.length; i++) {
        this.buffer.push(bytes[i]);
      }
    }
  }

  private writeBinary(data: Uint8Array): void {
    const len = data.length;
    if (len < 256) {
      this.buffer.push(TAG_BINARY_8, len);
    } else {
      this.buffer.push(
        TAG_BINARY_32,
        (len >> 24) & 0xff,
        (len >> 16) & 0xff,
        (len >> 8) & 0xff,
        len & 0xff
      );
    }
    for (let i = 0; i < data.length; i++) {
      this.buffer.push(data[i]);
    }
  }
}

File: crates/wazeko-core/src/protocol/encoder.rs
use crate::error::{ProtocolError, Result};
use crate::protocol::node::{NodeContent, ProtocolNode};
use crate::protocol::tokens::get_token_index;

pub const TAG_LIST_EMPTY: u8 = 0x00;
pub const TAG_LIST_8: u8 = 0x01;
pub const TAG_LIST_16: u8 = 0x02;
pub const TAG_BINARY_8: u8 = 0x03;
pub const TAG_BINARY_32: u8 = 0x04;
pub const TAG_STRING: u8 = 0x05;
pub const TAG_TOKEN: u8 = 0x06;

pub struct Encoder {
    buffer: Vec<u8>,
}

impl Default for Encoder {
    fn default() -> Self {
        Self::new()
    }
}

impl Encoder {
    pub fn new() -> Self {
        Self { buffer: Vec::new() }
    }

    pub fn encode(&mut self, node: &ProtocolNode) -> Result<Vec<u8>> {
        self.buffer.clear();
        self.write_node(node)?;
        Ok(self.buffer.clone())
    }

    fn write_node(&mut self, node: &ProtocolNode) -> Result<()> {
        let has_content = !matches!(node.content, NodeContent::Nil);
        let list_size = 1 + (node.attrs.len() * 2) + if has_content { 1 } else { 0 };

        self.write_list_header(list_size)?;
        self.write_string(&node.tag)?;

        for (k, v) in &node.attrs {
            self.write_string(k)?;
            self.write_string(v)?;
        }

        match &node.content {
            NodeContent::Nil => {}
            NodeContent::Binary(bytes) => {
                self.write_binary(bytes)?;
            }
            NodeContent::List(children) => {
                self.write_list_header(children.len())?;
                for child in children {
                    self.write_node(child)?;
                }
            }
        }

        Ok(())
    }

    fn write_list_header(&mut self, size: usize) -> Result<()> {
        if size == 0 {
            self.buffer.push(TAG_LIST_EMPTY);
        } else if size < 256 {
            self.buffer.push(TAG_LIST_8);
            self.buffer.push(size as u8);
        } else if size < 65536 {
            self.buffer.push(TAG_LIST_16);
            self.buffer.extend_from_slice(&(size as u16).to_be_bytes());
        } else {
            return Err(ProtocolError::EncodingError("List size exceeds 65535".into()).into());
        }
        Ok(())
    }

    fn write_string(&mut self, text: &str) -> Result<()> {
        if let Some(idx) = get_token_index(text) {
            self.buffer.push(TAG_TOKEN);
            self.buffer.push(idx);
        } else {
            let bytes = text.as_bytes();
            let len = bytes.len();
            if len > 65535 {
                return Err(ProtocolError::EncodingError("String too long".into()).into());
            }
            self.buffer.push(TAG_STRING);
            self.buffer.extend_from_slice(&(len as u16).to_be_bytes());
            self.buffer.extend_from_slice(bytes);
        }
        Ok(())
    }

    fn write_binary(&mut self, data: &[u8]) -> Result<()> {
        let len = data.len();
        if len < 256 {
            self.buffer.push(TAG_BINARY_8);
            self.buffer.push(len as u8);
        } else {
            self.buffer.push(TAG_BINARY_32);
            self.buffer.extend_from_slice(&(len as u32).to_be_bytes());
        }
        self.buffer.extend_from_slice(data);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::protocol::decoder::Decoder;

    #[test]
    fn test_encode_decode_roundtrip() {
        let node = ProtocolNode::with_tag("iq")
            .attr("id", "12345")
            .attr("type", "get")
            .attr("to", "s.whatsapp.net");

        let mut encoder = Encoder::new();
        let bytes = encoder.encode(&node).expect("Encoding should succeed");
        assert!(!bytes.is_empty());

        let decoded = Decoder::decode(&bytes).expect("Decoding should succeed");
        assert_eq!(decoded.tag, "iq");
        assert_eq!(decoded.get_attr("id"), Some("12345"));
        assert_eq!(decoded.get_attr("type"), Some("get"));
        assert_eq!(decoded.get_attr("to"), Some("s.whatsapp.net"));
    }

    #[test]
    fn test_nested_node_roundtrip() {
        let child = ProtocolNode::with_tag("ping");
        let parent = ProtocolNode::with_tag("iq")
            .attr("id", "ping-1")
            .content_nodes(vec![child]);

        let mut encoder = Encoder::new();
        let bytes = encoder.encode(&parent).expect("Encoding should succeed");

        let decoded = Decoder::decode(&bytes).expect("Decoding should succeed");
        assert_eq!(decoded.tag, "iq");
        let child_node = decoded.get_child("ping").expect("Must contain ping child");
        assert_eq!(child_node.tag, "ping");
    }
}

